beacon_node/network/src/sync/network_context.rs

```diff
@@ -405,4 +421,93 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
         self.range_blocks_and_blobs_requests
             .insert(id, (sender_id, info));
     }
+
+    // Request handlers
+
+    pub fn on_single_block_response(
+        &mut self,
+        request_id: SingleLookupReqId,
+        block: RpcEvent<Arc<SignedBeaconBlock<T::EthSpec>>>,
+    ) -> RpcProcessingResult<Arc<SignedBeaconBlock<T::EthSpec>>> {
+        let Entry::Occupied(mut request) = self.blocks_by_root_requests.entry(request_id) else {
+            return None;
+        };
+
+        Some(match block {
+            RpcEvent::Response(block, seen_timestamp) => {
+                match request.get_mut().add_response(block) {
+                    Ok(block) => Ok((block, seen_timestamp)),
+                    Err(e) => {
+                        // The request must be dropped after receiving an error.
+                        // TODO: We could NOT drop the request here, and penalize the peer again if
+                        // sends multiple penalizable chunks after the first invalid.
+                        request.remove();
+                        Err(e)
+                    }
+                }
+            }
+            RpcEvent::StreamTermination => match request.remove().terminate() {
```

Comment:
Do we need to handle the scenario where a peer doesn't send stream termination? (`ActiveBlocksByRootRequest` stored is quite small in size though)

Reply:
Current sync codebase expects the stream termination to always appear when syncing or doing blob lookups. The API contract with network's ReqResp behaviour is that you always get either an error or stream termination. There are timeouts in place in case the peer leaves the stream open and doesn't send the termination.